### PR TITLE
pytorch multi-gpu and fixing other things

### PR DESCRIPTION
**Pull request by vbvg2008**

### CODE REVIEW

File: fastestimator/network.py [hunk 1]
```diff
@@ -176,45 +154,82 @@ def __init__(self, ops: Iterable[Union[TensorOp, Scheduler[TensorOp]]]):
...
-    def forward_step_eager(self,
-                           batch: Dict[str, Any],
-                           state: Dict[str, Any],
-                           ops: List[TensorOp],
-                           effective_outputs: List[str]) -> Dict[str, Any]:
-        prediction = {}
+    def _get_effective_batch_input(self, batch: MutableMapping[str, Any], mode: str) -> Dict[str, Any]:
+        #copy data to gpu
+        if self.device.type == "cuda":
+            new_batch = {key: batch[key].to(self.device) for key in self.effective_inputs[mode] if key in batch}
```
[the remark this answers is not in the document]
Comment (vbvg2008) [reply fastestimator/network.py]: nope, this is copying the data from cpu to gpu.

[the remark this answers is not in the document]
Comment (vbvg2008) [reply fastestimator/network.py]: nope, it will send the data to one gpu first, then from one gpu broadcast to others.

File: fastestimator/network.py [hunk 2]
```diff
@@ -176,45 +154,82 @@ def __init__(self, ops: Iterable[Union[TensorOp, Scheduler[TensorOp]]]):
...
+        return new_batch
+
+    def run_step(self, batch: Dict[str, Any], state: Dict[str, Any]) -> Tuple[Dict[str, Any]]:
+        batch_in = self._get_effective_batch_input(batch, state["mode"])
         mode = state["mode"]
         state["tape"] = NonContext()
-        if self.device.type == "cuda":
-            for key, val in batch.items():
-                batch[key] = val.to(self.device)
+        #gpu operation
```
[the remark this answers is not in the document]
Comment (vbvg2008) [reply fastestimator/network.py]: fixed.

File: fastestimator/network.py [hunk 3]
```diff
@@ -176,45 +154,82 @@ def __init__(self, ops: Iterable[Union[TensorOp, Scheduler[TensorOp]]]):
...
-            if key in batch:
-                value = batch[key]
-                if self.device.type == "cuda":
-                    value = value.to("cpu")
-                prediction[key] = value
-        return prediction
+            self._forward_batch(batch_in, state, self.epoch_ops)
+        #copy data to cpu
+        if self.device.type == "cuda":
+            prediction = {key: batch_in[key].to("cpu") for key in self.effective_outputs[mode] if key in batch_in}
```
Comment (TortoiseHam) [inline fastestimator/network.py]: batch_in[key].detach().to("cpu")

to clear graph instead of making last backward() a retrain_graph=False

File: fastestimator/network.py [hunk 4]
```diff
@@ -176,45 +154,82 @@ def __init__(self, ops: Iterable[Union[TensorOp, Scheduler[TensorOp]]]):
...
 
     def load_epoch(self, mode: str, epoch: int) -> List[TensorOp]:
-        epoch_ops = super().load_epoch(mode, epoch)
+        super().load_epoch(mode, epoch)
         if self.device.type == "cuda":
             for model in self.epoch_models:
                 model.to(self.device)
-        return epoch_ops
+        update_ops = [op for op in self.epoch_ops if isinstance(op, UpdateOp)]
+        for idx, update_op in enumerate(update_ops):
```
Comment (TortoiseHam) [inline fastestimator/network.py]: we can drop this and use .detach() when collecting the tensors at the end instead

File: fastestimator/network.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (vbvg2008) [reply fastestimator/network.py]: yes, but copying data has never been an issue to GPU memory, because the size for one batch of data is very small comparing to size of network.
